File: tsvavg.py
#!/usr/bin/python

import sys


if len(sys.argv) < 2:
    print "usage:", sys.argv[0], " <tsv file>"
    print "prints a tsv row with the average values of  with only the fields provided on the command line"
    print "assumes the first line of the tab-separated values file is the header"
    print "assumes the input columns are all convertible to floats"
    print "<tsv file> may be specifed as '-' for stdin"
    exit()

tsvfilename = sys.argv[1]
if tsvfilename == '-':
    tsvfile = sys.stdin
else:
    tsvfile = open(sys.argv[1])

header = tsvfile.readline()
header_fields = header.strip().split('\t')
columns = [] # array of fields to keep, true if we keep false otherwise

count = {}
total = {}
for field in header_fields:
    count[field] = 0
    total[field] = 0

sys.stdout.write('\t'.join(header_fields) + '\n')

for line in tsvfile:
    fields = line.strip().split('\t')
    i = 0
    for field in header_fields:
        try:
            count[field] += float(fields[i])
            total[field] += 1
        except:
            continue
        i += 1

print '\t'.join([str(total[field] / count[field] if count[field] > 0 else 0) for field in header_fields])

File: tsvavg.cpp
#include <iostream>
#include <sstream>
#include <cstdlib>
#include <map>
#include <vector>
#include "join.h"
#include "split.h"

using namespace std;

int main(void) {

    string header;
    vector<string> header_fields;
    getline(cin, header);

    if (header.empty()) {
        cerr << "could not get header line" << endl;
        exit(1);
    }

    split(header, '\t', header_fields);
    map<string, unsigned int> counts;
    map<string, long double> totals;

    string line;
    getline(cin, line);
    while (!line.empty()) {
        vector<string> fields = split(line, '\t');
        int i = 0;
        for (vector<string>::iterator f = header_fields.begin(); f != header_fields.end(); ++f) {
            const string& field = *f;
            ++counts[field];
            totals[field] += atof(fields[i++].c_str());
        }
        getline(cin, line);
    }

    vector<string> averages;
    for (vector<string>::iterator f = header_fields.begin(); f != header_fields.end(); ++f) {
        const string& field = *f;
        if (counts[field] != 0) {
            stringstream s;
            s << totals[field] / counts[field];
            averages.push_back(s.str());
        }
    }

    cout << header << endl
         << join(averages, "\t") << endl;

    return 0;

}
